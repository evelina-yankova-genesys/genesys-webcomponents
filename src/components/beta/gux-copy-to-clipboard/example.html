<h1>gux-copy-to-clipboard</h1>

<span style="font-weight: bold">Email:</span>
<gux-copy-to-clipboard-beta>
  <div slot="content">jane.doe@example.com</div>
</gux-copy-to-clipboard-beta>

<br />

<span style="font-weight: bold">Phone Number:</span>
<gux-copy-to-clipboard-beta>
  <div slot="content">(123) 456-7890</div>
</gux-copy-to-clipboard-beta>
