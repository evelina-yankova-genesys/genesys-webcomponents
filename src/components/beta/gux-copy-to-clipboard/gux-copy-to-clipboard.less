@import (reference) '../../../style/spacing.less';
@import (reference) '../../../style/color-palette.less';
@import (reference) '../../../style/mixins.less';

button {
  all: unset;
}

.gux-copy-to-clipboard-wrapper {
  display: inline-flex;
  align-items: center;

  .gux-copy-content {
    display: inherit;
    align-items: inherit;
    cursor: pointer;

    ::slotted(*) {
      padding: 0 @spacing-3xs;
      margin-right: @spacing-2xs;
    }

    &:hover {
      .focus-and-hover-copy-content();
    }

    gux-icon[icon-name='copy'] {
      width: 16px;
      height: 16px;
      padding: @spacing-3xs;
      visibility: hidden;
      background-color: @gux-blue-100;
      border-radius: 4px;
    }
  }

  &:focus {
    outline: none;
    .focus-and-hover-copy-content();
  }

  gux-tooltip {
    display: inline-flex;
    align-items: center;

    gux-icon {
      width: 16px;
      height: 16px;
      padding-right: @spacing-xs;

      &[icon-name='badge-check'] {
        color: @gux-alert-green-60;
      }

      &[icon-name='badge-x'] {
        color: @gux-alert-red-60;
      }
    }
  }
}

.focus-and-hover-copy-content {
  ::slotted(*) {
    background-color: @gux-blue-100;
    border-radius: 4px;
  }

  gux-icon[icon-name='copy'] {
    visibility: visible;
  }
}
