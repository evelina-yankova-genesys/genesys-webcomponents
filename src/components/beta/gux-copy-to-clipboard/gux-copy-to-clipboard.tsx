import { Component, Element, h, JSX, State, Listen } from '@stencil/core';
import { trackComponent } from '../../../usage-tracking';
import { buildI18nForComponent, GetI18nValue } from '../../../i18n';
import translationResources from './i18n/en.json';

@Component({
  styleUrl: 'gux-copy-to-clipboard.less',
  tag: 'gux-copy-to-clipboard-beta',
  shadow: true
})
export class GuxCopyToClipboard {
  private i18n: GetI18nValue;

  @Element()
  private root: HTMLElement;

  @State()
  tooltipContent: string = 'offerClick';

  @Listen('mouseleave')
  onMouseleave() {
    this.resetTooltip();
  }

  @Listen('focusout')
  onFocusout() {
    this.resetTooltip();
  }

  @Listen('focus')
  onFocus() {
    // when element is focused by keyboard
    this.tooltipContent = 'offerEnter';
  }

  private resetTooltip() {
    this.tooltipContent = 'offerClick';
  }

  private onCopyToClipboard() {
    const copyText = this.root.innerText;

    navigator.clipboard
      .writeText(copyText)
      .then(() => {
        this.tooltipContent = 'success';
      })
      .catch(() => {
        this.tooltipContent = 'error';
      });
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);
    this.i18n = await buildI18nForComponent(this.root, translationResources);
  }

  render(): JSX.Element {
    return (
      <button
        onClick={this.onCopyToClipboard.bind(this)}
        class="gux-copy-to-clipboard-wrapper"
        title={this.i18n('copyToClipboard')}
      >
        <div class="gux-copy-content">
          <slot name="content" />
          <gux-icon icon-name="copy" decorative></gux-icon>
        </div>
        <gux-tooltip placement="bottom-end">
          {/* Icon */}
          {this.tooltipContent === 'success' ? (
            <gux-icon icon-name="badge-check" decorative></gux-icon>
          ) : (
            ''
          )}
          {this.tooltipContent === 'error' ? (
            <gux-icon icon-name="badge-x" decorative></gux-icon>
          ) : (
            ''
          )}

          {/* Tooltip Text */}
          {this.tooltipContent === 'offerClick' ? this.i18n('clickToCopy') : ''}
          {this.tooltipContent === 'offerEnter' ? this.i18n('enterToCopy') : ''}
          {this.tooltipContent === 'success' ? this.i18n('copySuccess') : ''}
          {this.tooltipContent === 'error' ? this.i18n('copyFailure') : ''}
        </gux-tooltip>
      </button>
    ) as JSX.Element;
  }
}
