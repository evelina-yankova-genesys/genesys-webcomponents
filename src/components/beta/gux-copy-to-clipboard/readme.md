# gux-copy-to-clipboard-beta



<!-- Auto Generated Below -->


## Dependencies

### Depends on

- [gux-icon](../../stable/gux-icon)
- [gux-tooltip](../../stable/gux-tooltip)

### Graph
```mermaid
graph TD;
  gux-copy-to-clipboard-beta --> gux-icon
  gux-copy-to-clipboard-beta --> gux-tooltip
  style gux-copy-to-clipboard-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
