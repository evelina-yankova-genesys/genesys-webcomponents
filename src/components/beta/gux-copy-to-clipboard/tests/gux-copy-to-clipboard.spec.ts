import { newSpecPage } from '@stencil/core/testing';
import * as popperjs from '@popperjs/core';
import MutationObserver from 'mutation-observer';
import { GuxCopyToClipboard } from '../gux-copy-to-clipboard';

const components = [GuxCopyToClipboard];
const language = 'en';

const html = `
  <gux-copy-to-clipboard-beta>
    <div slot="content">Test</div>
  </gux-copy-to-clipboard-beta>
`;

describe('gux-copy-to-clipboard-beta', () => {
  beforeEach(async () => {
    global.MutationObserver = MutationObserver;
    // popperjs does not work with Stencils MockHTMLElements used in tests
    jest.spyOn(popperjs, 'createPopper').mockReturnValue({
      destroy: jest.fn()
    } as unknown as popperjs.Instance);
  });

  afterEach(async () => {
    jest.spyOn(popperjs, 'createPopper').mockRestore();
  });

  describe('#render', () => {
    it('should render copy to clipboard', async () => {
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxCopyToClipboard);

      expect(page.root).toMatchSnapshot();
    });
  });

  describe('private', () => {
    it('should copy to clipboard successfully', async () => {
      const page = await newSpecPage({ components, html, language });
      const component = new GuxCopyToClipboard();

      Object.assign(window.navigator, {
        clipboard: {
          writeText: jest.fn().mockImplementation(() => Promise.resolve())
        }
      });

      component.onCopyToClipboard();
      await page.waitForChanges();

      expect(component.tooltipContent).toEqual('success');
      expect(window.navigator.clipboard.writeText).toHaveBeenCalled();
    });

    it('should copy to clipboard with error', async () => {
      const page = await newSpecPage({ components, html, language });
      const component = new GuxCopyToClipboard();

      Object.assign(window.navigator, {
        clipboard: {
          writeText: jest.fn().mockImplementation(() => Promise.reject())
        }
      });

      component.onCopyToClipboard();
      await page.waitForChanges();

      expect(component.tooltipContent).toEqual('error');
      expect(window.navigator.clipboard.writeText).toHaveBeenCalled();
    });
  });
});
