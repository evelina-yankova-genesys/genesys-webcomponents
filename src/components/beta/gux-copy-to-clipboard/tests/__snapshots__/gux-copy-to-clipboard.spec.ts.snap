// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-copy-to-clipboard-beta #render should render copy to clipboard 1`] = `
<gux-copy-to-clipboard-beta>
  <mock:shadow-root>
    <button class="gux-copy-to-clipboard-wrapper" title="Copy to clipboard">
      <div class="gux-copy-content">
        <slot name="content"></slot>
        <gux-icon decorative="" icon-name="copy"></gux-icon>
      </div>
      <gux-tooltip placement="bottom-end">
        Click to Copy
      </gux-tooltip>
    </button>
  </mock:shadow-root>
  <div slot="content">
    Test
  </div>
</gux-copy-to-clipboard-beta>
`;
